src/mqt/predictor/reward.py

```diff
@@ -70,3 +73,95 @@ def calc_qubit_index(qargs: list[Qubit], qregs: list[QuantumRegister], index: in
             return qubit_index
     error_msg = f"Global qubit index for local qubit {index} index not found."
     raise ValueError(error_msg)
+
+
+def expected_success_probability(qc: QuantumCircuit, device: Device, precision: int = 10) -> float:
+    """Calculates the expected success probability of a given quantum circuit on a given device.
+
+    It is calculated by multiplying the expected fidelity with a min(T1,T2)-dependant
+    decay factor during qubit idle times. Idle times are available after scheduling.
+
+    Arguments:
+        qc: The quantum circuit to be compiled.
+        device: The device to be used for compilation.
+        precision: The precision of the returned value. Defaults to 10.
+
+    Returns:
+        The expected success probability of the given quantum circuit on the given device.
+    """
+    # collect gate and measurement durations for active qubits
+    op_times, active_qubits = [], set()
+    try:
+        for instruction, qargs, _cargs in qc.data:
+            gate_type = instruction.name
+
+            if gate_type == "barrier" or gate_type == "id":
+                continue
+            assert len(qargs) in [1, 2]
+            first_qubit_idx = calc_qubit_index(qargs, qc.qregs, 0)
+            active_qubits.add(first_qubit_idx)
+
+            if len(qargs) == 1:  # single-qubit gate
+                if gate_type == "measure":
+                    duration = device.get_readout_duration(first_qubit_idx)
+                else:
+                    duration = device.get_single_qubit_gate_duration(gate_type, first_qubit_idx)
+                op_times.append((gate_type, [first_qubit_idx], duration, "s"))
+            else:  # multi-qubit gate
+                second_qubit_idx = calc_qubit_index(qargs, qc.qregs, 1)
+                active_qubits.add(second_qubit_idx)
+                duration = device.get_two_qubit_gate_duration(gate_type, first_qubit_idx, second_qubit_idx)
+                op_times.append((gate_type, [first_qubit_idx, second_qubit_idx], duration, "s"))
+    except ValueError as err:
```

Comment:
Do you mean sth. like this @burgholzer? Shouldn't this rather be part of MQT Bench then?
```
def esp_calibration_data_is_available(device: Device) -> bool:
    """Check if calibration data is available for the device."""
    
    assert device.calibration is not None
    for qubit in range(device.num_qubits):
        assert qubit in device.calibration.single_qubit_gate_fidelity
        for gate in device.get_single_qubit_gates():
            assert gate in device.calibration.single_qubit_gate_fidelity[qubit]
            assert device.calibration.single_qubit_gate_fidelity[qubit][gate] > 0
        assert qubit in device.calibration.readout_fidelity

    for qubit1, qubit2 in device.coupling_map:
        assert (qubit1, qubit2) in device.calibration.two_qubit_gate_fidelity
        for gate in device.get_two_qubit_gates():
            assert gate in device.calibration.two_qubit_gate_fidelity[qubit1, qubit2]
            assert device.calibration.two_qubit_gate_fidelity[qubit1, qubit2][gate] > 
```

Reply:
Yeah, this is precisely along the lines of what I had in mind.
Such a function would check the availability of all the necessary data at once and, then, you would not need `try...catch` blocks at all in the actual library code.

Given that MQT Bench has no notion of the ESP so far, I would have rather placed this here.
Although I could see a scenario where it makes sense to have the figures of merit and their calculation already in MQT Bench (given that it's a benchmarking library after all).
In order to get this PR merged, I'd probably place it here first and add it to MQT Bench later. Once it is integrated in MQT Bench, it can be removed here (while potentially reexporting the function from MQT Bench).